### ISSUE TITLE
Can't dynamically map type from class private field

### ISSUE THREAD

**Issue by k3nsei** [microsoft/TypeScript#44633]
# Suggestion

Support private fields in mapped types.

## 🔍 Search Terms

typescript, private field, mapped types

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [ ] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Include information about private fields types inside class interface.

## 📃 Motivating Example

TypeScript have dynamic mapped types which is very powerful feature. But currently can't used with private fields. So its hard to migrate to those.

## 💻 Use Cases

**Working example without private field:**
```typescript
interface Consumer<T> {
    (data: T): void;
}

class EventEmitter<T> {
    emit(data: T): void {
        return void 0;
    }

    subscribe(consumer: Consumer<T>): () => void {
        const unsubscribe = () => void 0;

        return unsubscribe;
    }
}

class Host {
    private readonly events = {
        init: new EventEmitter<Record<string, number>>(),
        contextUpdate: new EventEmitter<Record<string, string>>(),
        inactivity: new EventEmitter<boolean>(),
    } as const;

    on<T extends HostEventType>(
        type: T,
        consumer: HostEventConsumer<T>,
    ): () => void {
        return this.events[type].subscribe(consumer);
    }
}

type HostEventType = keyof Host['events'];

type HostEventConsumer<T extends HostEventType> =
    Host['events'][T] extends EventEmitter<infer U> ? Consumer<U> : never;

const host = new Host();

// Here data is mapped to Record<string, number>
host.on('init', (data) => console.log({ data }));
```

**Not working example with private field (expecting to work in future):**
```typescript
interface Consumer<T> {
    (data: T): void;
}

class EventEmitter<T> {
    emit(data: T): void {
        return void 0;
    }

    subscribe(consumer: Consumer<T>): () => void {
        const unsubscribe = () => void 0;

        return unsubscribe;
    }
}

class Host {
    readonly #events = {
        init: new EventEmitter<Record<string, number>>(),
        contextUpdate: new EventEmitter<Record<string, string>>(),
        inactivity: new EventEmitter<boolean>(),
    } as const;

    on<T extends HostEventType>(
        type: T,
        consumer: HostEventConsumer<T>,
    ): () => void {
        return this.#events[type].subscribe(consumer);
    }
}

type HostEventType = keyof Host['#events'];

type HostEventConsumer<T extends HostEventType> =
    Host['#events'][T] extends EventEmitter<infer U> ? Consumer<U> : never;

const host = new Host();

// Here data should be mapped to Record<string, number>
host.on('init', (data) => console.log({ data }));
```

**Comment by nmain** [comment]
> `Host['#events']`

But that syntax already has a meaning; it's referring to a plain (not runtime-private) property with name `'#events'`:

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEoFcQjgIwDYoDeAUMmcgBQCUAXMgG4D2wAJgNzEC+xxoksiFAGFGIAM7oAttAA8AFQB8yEuUos4YOHTm0Gzdlx4I8cMWOQBRehHAXJwMP3lKV5CPbAV1m7bqatlUlVyKAgwdCgQPQCABg5VbiCyCRwxBChgHAgKBFEJaSg6EXEpWUVdDCxcAkDg8lzxMGRMFLSMrOQAXkoqLqV-FmQ4njqyUPDI5pLU9MyIePJuRONTcwAJRjEm1zGIOBZRPABPZAAiAGIIa3AxU67a0dAHOhAIAHdLa7A7BydUCFyUBYMi2GRAAHMADTIEBSLJQBQKaiQpLBBqQAAeYAAqgAHbwQF7vT42b4eP4AxhAkFgMFQ5Cg0DgxHI1GqUCIMDAegOI5Ej5WUk-RyyHCMRgEOAgJFUFEJZCmZANLYcVGieTICBYmwsdabb5fORHXEQJFs8hgY2E5ByOWjZWlQrIDZbQXgYr5MoKO1kCqYbD4Ijm3YTKJgAAWwDEAG0LldSbcALrRy0mxMAOlasyyOTyjqoCzISx4qZQLoNpKNJvuAGsIEdGDBnfrowByS5fMStxOq0vN11fD2OjVayAgXX9ivgKumrqo8ttjsJ7vRuSJzXa8fmN1k36yUAwaDIbFKAD8yCHBRkJ+QROsUFVyqa4f191eH3L1FVAHpv87oCg3hwAyL7oHggwdJIcC4iagxgIwaCUtSjIQtCsKSPCCjEC+WzpqIFCtk8YCttCXgaHAvSdEoyoShA6Z4Iw4IUIQyBAcgnBUAWQA)

**Comment by pstepniewskiconfigit** [comment]
@nmain `"#events"` is not private field.

**Comment by k3nsei** [comment]
@nmain "#events" is not private field.

![image](https://user-images.githubusercontent.com/190422/122409804-b89e9c80-cf83-11eb-93bc-0aadd03287c6.png)

**Comment by nmain** [comment]
> @nmain `"#events"` in your example is not real private field.

Yes, I'm well aware of that.  That was my point; you can't use `Host['#events']` as syntax in typescript types for a private field because it already means something else.

**Comment by RyanCavanaugh** [comment]
It's pretty hard for me to see us doing this. Indexing into a `private` field like shown here is already quite fraught -- because types are stripped from `private` fields in `.d.ts` emit, it becomes effectively `any` in any downstream compilation, which is super bad. The motivation to derive the types from the initializer value here is noble, but combined with `private` it's just not a great scenario and only seems worse with `#` fields.

**Comment by pstepniewskiconfigit** [comment]
@RyanCavanaugh I only need that only during writing code as a type guard not as something that will ends up in `.d.ts` used by some consumers. Private fields are used internally anyways.

**Comment by pstepniewskiconfigit** [comment]
@RyanCavanaugh I know this is problematic. There are many disadvantages to using private fields with TypeScript these days.
